src/content/symbol-store.js

```diff
@@ -14,70 +32,100 @@ export class SymbolStore {
-      return Promise.reject(new Error('We\'ve tried to request a symbol table for this library before and failed, so we\'re not trying again.'));
+      return Promise.reject(
+        new Error('We\'ve tried to request a symbol table for this library before and failed, so we\'re not trying again.'));
+    }
+
+    const cachedTable = this._cachedSymbolTables.get(libid);
+    if (cachedTable !== undefined) {
+      return Promise.resolve(cachedTable);
     }
-    return this._db.getLibKey(debugName, breakpadId).catch(() => {
-      if (this._importingLibs.has(libid)) {
-        // We've already requested a symbol table for this library and are
-        // waiting for the result, so just return the promise for the existing
-        // request.
-        return this._importingLibs.get(libid);
-      }
 
+    const existingRequest = this._requestedSymbolTables.get(libid);
+    if (existingRequest !== undefined) {
+      // We've already requested a symbol table for this library and are
+      // waiting for the result, so just return the promise for the existing
+      // request.
+      return existingRequest;
+    }
+
+    // Try to get the symbol table from the database
+    const symbolTablePromise = this._db.getSymbolTable(debugName, breakpadId).catch(() => {
       // Request the symbol table from the symbol provider.
-      const symbolTablePromise = this._symbolProvider.requestSymbolTable(lib.debugName, lib.breakpadId);
-      const importPromise = symbolTablePromise.then(symbolTable => {
-        this._importingLibs.delete(libid);
-        return this._db.importLibrary(debugName, breakpadId, symbolTable);
-      }, error => {
-        this._importingLibs.delete(libid);
+      const symbolTablePromise = this._symbolProvider.requestSymbolTable(debugName, breakpadId).catch(error => {
         console.error(`Failed to symbolicate library ${debugName}: ${error}`);
+        this._failedRequests.add(libid);
+        this._requestedSymbolTables.delete(libid);
         throw error;
       });
-      this._importingLibs.set(libid, importPromise);
-      return importPromise;
-    }).catch(error => {
-      this._failedRequests.add(libid);
-      throw error;
+
+      // Once the symbol table comes in, store it in the database, but don't
+      // let that block the promise that we return to our caller.
+      symbolTablePromise.then(({addrs, index, buffer}) => {
+        this._db.storeSymbolTable(debugName, breakpadId, [addrs, index, buffer]).then(() => {
+          this._requestedSymbolTables.delete(libid);
+        }, error => {
+          console.error(`Failed to store the symbol table for ${debugName} ${breakpadId} in the database:`, error);
+        });
+      });
+
+      return symbolTablePromise;
+    });
+    this._requestedSymbolTables.set(libid, symbolTablePromise);
+    symbolTablePromise.then(symbolTable => {
+      this._cachedSymbolTables.set(libid, symbolTable);
     });
+    return symbolTablePromise;
   }
```

Comment:
nit: maybe use this opportunity to use async/await and make this method more readable